From 2cc5acef0829d5df81fa7169005a201fef1f9efd Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Mon, 8 Sep 2014 13:18:29 -0400
Subject: [ethernet] r8169: Enable RX_MULTI_EN for RTL_GIGA_MAC_VER_40

Message-id: <1410182312-19560-11-git-send-email-ivecera@redhat.com>
Patchwork-id: 91985
O-Subject: [RHEL7.1 PATCH 10/13] r8169: Enable RX_MULTI_EN for RTL_GIGA_MAC_VER_40
Bugzilla: 1139247
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: John Linville <linville@redhat.com>

BZ: #1139247

Upstream commit(s):
commit 7a9810e7bd99c922d9cedf64dbaa5ef6be412295
Author: Michel Danzer <michel@daenzer.net>
Date:   Thu Jul 17 12:55:40 2014 +0900

    r8169: Enable RX_MULTI_EN for RTL_GIGA_MAC_VER_40

    The ethernet port on my ASUS A88X Pro mainboard stopped working
    several times a day, with messages like these in dmesg:

    AMD-Vi: Event logged [IO_PAGE_FAULT device=05:00.0 domain=0x001e address=0x0000000000003000 flags=0x0050]

    Searching the web for these messages led me to similar reports about
    different hardware supported by r8169, and eventually to commits
    3ced8c955e74d319f3e3997f7169c79d524dfd06 ('r8169: enforce RX_MULTI_EN
    for the 8168f.') and eb2dc35d99028b698cdedba4f5522bc43e576bd2 ('r8169:
    RxConfig hack for the 8168evl'). So I tried this change, and it fixes
    the problem for me.

    Signed-off-by: Michel Danzer <michel@daenzer.net>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/realtek/r8169.c b/drivers/net/ethernet/realtek/r8169.c
index a0eb198..b0e5096 100644
--- a/drivers/net/ethernet/realtek/r8169.c
+++ b/drivers/net/ethernet/realtek/r8169.c
@@ -4226,6 +4226,8 @@ static void rtl_init_rxcfg(struct rtl8169_private *tp)
   RTL_W32(RxConfig, RX128_INT_EN | RX_MULTI_EN | RX_DMA_BURST);
   break;
  case RTL_GIGA_MAC_VER_40:
+  RTL_W32(RxConfig, RX128_INT_EN | RX_MULTI_EN | RX_DMA_BURST | RX_EARLY_OFF);
+  break;
  case RTL_GIGA_MAC_VER_41:
  case RTL_GIGA_MAC_VER_42:
  case RTL_GIGA_MAC_VER_43:
-- 
1.7.1